docs/ci_set_socket_functions.3
.\" Copyright (C) Daniel Stenberg
.\" SPDX-License-Identifier: MIT
.TH CI_SET_SOCKET_FUNCTIONS 3 "8 Oct 2024"
.SH NAME
ci_set_socket_functions, ci_set_socket_functions_ex \- Set socket io callbacks
.SH SYNOPSIS
.nf
#include <ci.h>

typedef enum {
  CI_SOCKFUNC_FLAG_NONBLOCKING = 1 << 0
} ci_sockfunc_flags_t;

typedef enum {
  CI_SOCKET_OPT_SENDBUF_SIZE,
  CI_SOCKET_OPT_RECVBUF_SIZE,
  CI_SOCKET_OPT_BIND_DEVICE,
  CI_SOCKET_OPT_TCP_FASTOPEN
} ci_socket_opt_t;

typedef enum {
  CI_SOCKET_CONN_TCP_FASTOPEN = 1 << 0
} ci_socket_connect_flags_t;

typedef enum {
  CI_SOCKET_BIND_TCP = 1 << 0,
  CI_SOCKET_BIND_CLIENT = 1 << 1
} ci_socket_bind_flags_t;

struct ci_socket_functions_ex {
  unsigned int version; /* ABI Version: must be "1" */
  unsigned int flags;

  ci_socket_t (*asocket)(int domain, int type, int protocol, void *user_data);
  int (*aclose)(ci_socket_t sock, void *user_data);
  int (*asetsockopt)(ci_socket_t sock, ci_socket_opt_t opt, const void *val,
                     ci_socklen_t val_size, void *user_data);
  int (*aconnect)(ci_socket_t sock, const struct sockaddr *address,
                  ci_socklen_t address_len, unsigned int flags,
                  void *user_data);
  ci_ssize_t (*arecvfrom)(ci_socket_t sock, void *buffer, size_t length,
                            int flags, struct sockaddr *address,
                            ci_socklen_t *address_len, void *user_data);
  ci_ssize_t (*asendto)(ci_socket_t sock, const void *buffer, size_t length,
                          int flags, const struct sockaddr *address,
                          ci_socklen_t address_len, void *user_data);
  int (*agetsockname)(ci_socket_t sock, struct sockaddr *address,
                      ci_socklen_t *address_len, void *user_data);
  int (*abind)(ci_socket_t sock, unsigned int flags,
               const struct sockaddr *address, socklen_t address_len,
               void *user_data);
  unsigned int (*aif_nametoindex)(const char *ifname, void *user_data);
  const char *(*aif_indextoname)(unsigned int ifindex, char *ifname_buf,
                                 size_t ifname_buf_len, void *user_data);
};

ci_status_t ci_set_socket_functions_ex(ci_channel_t *channel,
  const struct ci_socket_functions_ex *funcs, void *user_data);


struct ci_socket_functions {
    ci_socket_t (*\fIasocket\fP)(int, int, int, void *);
    int (*\fIaclose\fP)(ci_socket_t, void *);
    int (*\fIaconnect\fP)(ci_socket_t, const struct sockaddr *, ci_socklen_t, void *);
    ci_ssize_t (*\fIarecvfrom\fP)(ci_socket_t, void *, size_t, int,
                              struct sockaddr *, ci_socklen_t *, void *);
    ci_ssize_t (*\fIasendv\fP)(ci_socket_t, const struct iovec *, int, void *);
};

void ci_set_socket_functions(ci_channel_t *\fIchannel\fP,
                               const struct ci_socket_functions * \fIfunctions\fP,
                               void *\fIuser_data\fP);
.fi
.SH DESCRIPTION
.PP

\fBci_set_socket_functions_ex(3)\fP sets a set of callback \fIfunctions\fP in
the given ci channel handle.  Cannot be used when \fBCI_OPT_EVENT_THREAD\fP
is passed to \fIci_init_options(3)\fP.  This function replaces the now
deprecated \fBci_set_socket_functions(3)\fP call.

These callback functions will be invoked to create/destroy socket objects and
perform io, instead of the normal system calls. A client application can
override normal network operation fully through this functionality, and provide
its own transport layer.

Some callbacks may be optional and are documented as such below, but failing
to implement such callbacks will disable certain features within c-ci.  It
is strongly recommended to implement all callbacks.

All callback functions are expected to operate like their system equivalents,
and to set \fBerrno(2)\fP or \fBWSASetLastError(2)\fP to an appropriate error
code on failure. It is strongly recommended that io callbacks are implemented
to be asynchronous and indicated as such in the \fIflags\fP member.  The io
callbacks can return error codes of \fBEAGAIN\fP, \fBEWOULDBLOCK\fP, or
\fBWSAEWOULDBLOCK\fP when they would otherwise block.

The \fIuser_data\fP value is provided to each callback function invocation to
serve as context.

The \fBci_set_socket_functions_ex(3)\fP must provide the following structure
members and callbacks (which are different from the
\fBci_set_socket_functions(3)\fP members and callbacks):

.RS 4
.TP 8
.B unsigned int \fIversion\fP
.br
ABI Version of structure.  Must be set to a value of "1".

.TP 8
.B unsigned int \fIflags\fP
.br
Flags available are specified in \fIci_sockfunc_flags_t\fP.

.TP 8
.B ci_socket_t (*\fIasocket\fP)(int \fIdomain\fP, int \fItype\fP, int \fIprotocol\fP, void * \fIuser_data\fP)
.br
\fIREQUIRED\fP. Creates an endpoint for communication and returns a descriptor. \fIdomain\fP,
\fItype\fP, and \fIprotocol\fP each correspond to the parameters of
\fBsocket(2)\fP. Returns a handle to the newly created socket, or
\fBCI_SOCKET_BAD\fP on error.

.TP 8
.B int (*\fIaclose\fP)(ci_socket_t \fIfd\fP, void * \fIuser_data\fP)
.br
\fIREQUIRED\fP. Closes the socket endpoint indicated by \fIfd\fP. See \fBclose(2)\fP.

.TP 8
.B int (*\fIasetsockopt\fP)(ci_socket_t \fIfd\fP, ci_socket_opt_t \fIopt\fP, const void * \fIval\fP, ci_socklen_t \fIval_size\fP, void * \fIuser_data\fP)
.br
\fIREQUIRED\fP. Set socket option.  This shci a similar syntax to the BSD \fIsetsockopt(2)\fP
call, however c-ci uses different options for portability. The value is
a pointer to the desired value, and each option has its own data type listed
in the options below defined in \fIci_socket_opt_t\fP.

.TP 8
.B int (*\fIaconnect\fP)(ci_socket_t \fIfd\fP, const struct sockaddr * \fIaddr\fP, ci_socklen_t \fIaddr_len\fP, unsigned int \fIflags\fP, void * \fIuser_data\fP)
.br
\fIREQUIRED\fP. Initiate a connection to the address indicated by \fIaddr\fP on
a socket. Additional flags controlling behavior are in
\fIci_socket_connect_flags_t\fP. See \fBconnect(2)\fP.

.TP 8
.B ci_ssize_t (*\fIarecvfrom\fP)(ci_socket_t \fIfd\fP, void * \fIbuffer\fP, size_t \fIbuf_size\fP, int \fIflags\fP, struct sockaddr * \fIaddr\fP, ci_socklen_t * \fIaddr_len\fP, void * \fIuser_data\fP)
.br
\fIREQUIRED\fP. Receives data from remote socket endpoint, if available. If the
\fIaddr\fP parameter is not NULL and the connection protocol provides the source
address, the callback should fill this in. The \fIflags\fP parameter is
currently unused. See \fBrecvfrom(2)\fP.

.TP 8
.B ci_ssize_t (*\fIasendto\fP)(ci_socket_t \fIfd\fP, const void * \fIbuffer\fP, size_t \fIlength\fP, int \fIflags\fP, const struct sockaddr * \fIaddress\fP, ci_socklen_t \fIaddress_len\fP, void * \fIuser_data\fP)
.br
\fIREQUIRED\fP. Send data, as provided by the \fIbuffer\fP, to the socket
endpoint. The \fIflags\fP member may be used on systems that have
\fBMSG_NOSIGNAL\fP defined but is otherwise unused.  An \fIaddress\fP is
provided primarily to support TCP FastOpen scenarios, which will be NULL in
other circumstances. See \fBsendto(2)\fP.

.TP 8
.B int (*\fIagetsockname\fP)(ci_socket_t \fIfd\fP, struct sockaddr * \fIaddress\fP, ci_socklen_t * \fIaddress_len\fP, void * \fIuser_data\fP)
.br
\fIOptional\fP. Retrieve the local address of a socket and store it into the provided
\fIaddress\fP buffer. May impact DNS Cookies if not provided. See
\fBgetsockname(2)\fP.

.TP 8
.B int (*\fIabind\fP)(ci_socket_t \fIfd\fP, unsigned int \fIflags\fP, const struct sockaddr * \fIaddress\fP, ci_socklen_t \fIaddress_len\fP, void * \fIuser_data\fP)
.br
\fIOptional\fP. Bind the socket to an address.  This can be used for client
connections to bind the source address for packets before connect, or
for server connections to bind to an address and port before listening.
Currently c-ci only supports client connections.  \fIflags\fP from
\fIci_socket_bind_flags_t\fP can be specified.  See \fBbind(2)\fP.

.TP 8
.B unsigned int (*\fIaif_nametoindex\fP)(const char * \fIifname\fP, void * \fIuser_data\fP)
.br
\fIOptional\fP. Convert an interface name into the interface index.  If this
callback is not specified, then IPv6 Link-Local DNS servers cannot be used.
See \fBif_nametoindex(2)\fP.

.TP 8
.B const char * (*\fIaif_indextoname\fP)(unsigned int \fIifindex\fP, char * \fIifname_buf\fP, size_t \fIifname_buf_len\fP, void * \fIuser_data\fP)
.br
\fIOptional\fP. Convert an interface index into the interface name.  If this
callback is not specified, then IPv6 Link-Local DNS servers cannot be used.
\fIifname_buf\fP must be at least \fBIF_NAMESIZE\fP or \fBIFNAMSIZ\fP in size.
See \fBif_indextoname(2)\fP.
.RE

.PP
\fBci_sockfunc_flags_t\fP values:

.RS 4
.TP 8
.B \fICI_SOCKFUNC_FLAG_NONBLOCKING\fP
.br
Used to indicate the implementation of the io functions are asynchronous.
.RE

.PP
\fBci_socket_opt_t\fP values:

.RS 4
.TP 8
.B \fICI_SOCKET_OPT_SENDBUF_SIZE\fP
.br
Set the Send Buffer size.  Value is a pointer to an int. (SO_SNDBUF).

.TP 8
.B \fICI_SOCKET_OPT_RECVBUF_SIZE\fP
.br
Set the Receive Buffer size.  Value is a pointer to an int. (SO_RCVBUF).

.TP 8
.B \fICI_SOCKET_OPT_BIND_DEVICE\fP
.br
Set the network interface to use as the source for communication. Value is a C
string. (SO_BINDTODEVICE)

.TP 8
.B \fICI_SOCKET_OPT_TCP_FASTOPEN\fP
.br
Enable TCP Fast Open.  Value is a pointer to an \fIci_bool_t\fP.  On some
systems this could be a no-op if it is known it is on by default and
return success.  Other systems may be a no-op if known the system does
not support the feature and returns failure with errno set to \fBENOSYS\fP or
\fBWSASetLastError(WSAEOPNOTSUPP);\fP.
.RE

.PP
\fBci_socket_connect_flags_t\fP values:
.RS 4
.TP 8
.B \fICI_SOCKET_CONN_TCP_FASTOPEN\fP
.br
Connect using TCP Fast Open.
.RE

.PP
\fBci_socket_bind_flags_t\fP values:

.RS 4
.TP 8
.B \fICI_SOCKET_BIND_TCP\fP
.br
Bind is for a TCP connection.

.TP 19
.B \fICI_SOCKET_BIND_CLIENT\fP
.br
Bind is for a client connection, not server.
.RE

.PP

\fBci_set_socket_functions(3)\fP sets a set of callback \fIfunctions\fP in the
given ci channel handle.  Cannot be used when \fBCI_OPT_EVENT_THREAD\fP is
passed to \fIci_init_options(3)\fP.  This function is deprecated as of
c-ci 1.34.0 in favor of \fIci_set_socket_functions_ex(3)\fP.

\fBci_set_socket_functions(3)\fP allows you to choose to only implement
some of the socket functions, and provide NULL to any others and c-ci will use
its built-in system functions in that case.

.PP
All callback functions are expected to operate like their system equivalents,
and to set \fBerrno(2)\fP or \fBWSASetLastError(2)\fP to an appropriate error

RELEASE-NOTES.md
## c-ci version 1.34.2 - October 15 2024

This release contains a fix for downstream packages detecting the c-ci
version based on the contents of the header file rather than the
distributed pkgconf or cmake files.

## c-ci version 1.34.1 - October 9 2024

This release fixes a packaging issue.


## c-ci version 1.34.0 - October 9 2024

This is a feature and bugfix release.

Features:
* adig: read arguments from adigrc.
  [PR #856](https://github.com/c-ci/c-ci/pull/856)
* Add new pending write callback optimization via `ci_set_pending_write_cb`.
  [PR #857](https://github.com/c-ci/c-ci/pull/857)
* New function `ci_process_fds()`.
  [PR #875](https://github.com/c-ci/c-ci/pull/875)
* Failed servers should be probed rather than redirecting queries which could
  cause unexpected latency.
  [PR #877](https://github.com/c-ci/c-ci/pull/877)
* adig: rework command line arguments to mimic dig from bind.
  [PR #890](https://github.com/c-ci/c-ci/pull/890)
* Add new method for overriding network functions
  `ci_set_socket_function_ex()` to properly support all new functionality.
  [PR #894](https://github.com/c-ci/c-ci/pull/894)
* Fix regression with custom socket callbacks due to DNS cookie support.
  [PR #895](https://github.com/c-ci/c-ci/pull/895)
* ci_socket: set IP_BIND_ADDRESS_NO_PORT on ci_set_local_ip* tcp sockets
  [PR #887](https://github.com/c-ci/c-ci/pull/887)
* URI parser/writer for ci_set_servers_csv()/ci_get_servers_csv().
  [PR #882](https://github.com/c-ci/c-ci/pull/882)

Changes:
* Connection handling modularization.
  [PR #857](https://github.com/c-ci/c-ci/pull/857),
  [PR #876](https://github.com/c-ci/c-ci/pull/876)
* Expose library/utility functions to tools.
  [PR #860](https://github.com/c-ci/c-ci/pull/860)
* Remove `ci__` prefix, just use `ci_` for internal functions.
  [PR #872](https://github.com/c-ci/c-ci/pull/872)


Bugfixes:
* fix: potential WIN32_LEAN_AND_MEAN redefinition.
  [PR #869](https://github.com/c-ci/c-ci/pull/869)
* Fix googletest v1.15 compatibility.
  [PR #874](https://github.com/c-ci/c-ci/pull/874)
* Fix pkgconfig thread dependencies.
  [PR #884](https://github.com/c-ci/c-ci/pull/884)


Thanks go to these friendly people for their efforts and contributions for this
release:

* Brad House (@bradh352)
* Cristian Rodríguez (@crrodriguez)
* Georg (@tacerus)
* @lifenjoiner
* Shelley Vohr (@codebytere)
* 前进，前进，进 (@leleliu008)



src/lib/ci_set_socket_functions.c
/* MIT License
 *
 * Copyright (c) 2024 Brad House
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */
#include "ci_private.h"
#ifdef HAVE_SYS_UIO_H
#  include <sys/uio.h>
#endif
#ifdef HAVE_NETINET_IN_H
#  include <netinet/in.h>
#endif
#ifdef HAVE_NETINET_TCP_H
#  include <netinet/tcp.h>
#endif
#ifdef HAVE_NETDB_H
#  include <netdb.h>
#endif
#ifdef HAVE_ARPA_INET_H
#  include <arpa/inet.h>
#endif

#ifdef HAVE_STRINGS_H
#  include <strings.h>
#endif
#ifdef HAVE_SYS_IOCTL_H
#  include <sys/ioctl.h>
#endif
#ifdef NETWARE
#  include <sys/filio.h>
#endif

#include <assert.h>
#include <fcntl.h>
#include <limits.h>


#if defined(__linux__) && defined(TCP_FASTOPEN_CONNECT)
#  define TFO_SUPPORTED      1
#  define TFO_SKIP_CONNECT   0
#  define TFO_USE_SENDTO     0
#  define TFO_USE_CONNECTX   0
#  define TFO_CLIENT_SOCKOPT TCP_FASTOPEN_CONNECT
#elif defined(__FreeBSD__) && defined(TCP_FASTOPEN)
#  define TFO_SUPPORTED      1
#  define TFO_SKIP_CONNECT   1
#  define TFO_USE_SENDTO     1
#  define TFO_USE_CONNECTX   0
#  define TFO_CLIENT_SOCKOPT TCP_FASTOPEN
#elif defined(__APPLE__) && defined(HAVE_CONNECTX)
#  define TFO_SUPPORTED    1
#  define TFO_SKIP_CONNECT 0
#  define TFO_USE_SENDTO   0
#  define TFO_USE_CONNECTX 1
#  undef TFO_CLIENT_SOCKOPT
#else
#  define TFO_SUPPORTED 0
#endif

#ifndef HAVE_WRITEV
/* Structure for scatter/gather I/O. */
struct iovec {
  void  *iov_base; /* Pointer to data. */
  size_t iov_len;  /* Length of data.  */
};
#endif

ci_status_t
  ci_set_socket_functions_ex(ci_channel_t                        *channel,
                               const struct ci_socket_functions_ex *funcs,
                               void                                  *user_data)
{
  unsigned int known_versions[] = { 1 };
  size_t       i;

  if (channel == NULL || funcs == NULL) {
    return CI_EFORMERR;
  }

  /* Check to see if we know the version referenced */
  for (i = 0; i < sizeof(known_versions) / sizeof(*known_versions); i++) {
    if (funcs->version == known_versions[i]) {
      break;
    }
  }
  if (i == sizeof(known_versions) / sizeof(*known_versions)) {
    return CI_EFORMERR;
  }

  memset(&channel->sock_funcs, 0, sizeof(channel->sock_funcs));

  /* Copy individually for ABI compliance.  memcpy() with a sizeof would do
   * invalid reads */
  if (funcs->version >= 1) {
    if (funcs->asocket == NULL || funcs->aclose == NULL ||
        funcs->asetsockopt == NULL || funcs->aconnect == NULL ||
        funcs->arecvfrom == NULL || funcs->asendto == NULL) {
      return CI_EFORMERR;
    }
    channel->sock_funcs.version      = funcs->version;
    channel->sock_funcs.flags        = funcs->flags;
    channel->sock_funcs.asocket      = funcs->asocket;
    channel->sock_funcs.aclose       = funcs->aclose;
    channel->sock_funcs.asetsockopt  = funcs->asetsockopt;
    channel->sock_funcs.aconnect     = funcs->aconnect;
    channel->sock_funcs.arecvfrom    = funcs->arecvfrom;
    channel->sock_funcs.asendto      = funcs->asendto;
    channel->sock_funcs.agetsockname = funcs->agetsockname;
    channel->sock_funcs.abind        = funcs->abind;
  }

  /* Implement newer versions here ...*/


  channel->sock_func_cb_data = user_data;

  return CI_SUCCESS;
}

static int setsocknonblock(ci_socket_t sockfd, /* operate on this */
                           int           nonblock /* TRUE or FALSE */)
{
#if defined(HAVE_FCNTL_O_NONBLOCK)

  /* most recent unix versions */
  int flags;
  flags = fcntl(sockfd, F_GETFL, 0);
  if (nonblock) {
    return fcntl(sockfd, F_SETFL, flags | O_NONBLOCK);
  } else {
    return fcntl(sockfd, F_SETFL, flags & (~O_NONBLOCK)); /* LCOV_EXCL_LINE */
  }

#elif defined(HAVE_IOCTL_FIONBIO)

  /* older unix versions */
  int flags = nonblock ? 1 : 0;
  return ioctl(sockfd, FIONBIO, &flags);

#elif defined(HAVE_IOCTLSOCKET_FIONBIO)

#  ifdef WATT32
  char flags = nonblock ? 1 : 0;
#  else
  /* Windows */
  unsigned long flags = nonblock ? 1UL : 0UL;
#  endif
  return ioctlsocket(sockfd, (long)FIONBIO, &flags);

#elif defined(HAVE_IOCTLSOCKET_CAMEL_FIONBIO)

  /* Amiga */
  long flags = nonblock ? 1L : 0L;
  return IoctlSocket(sockfd, FIONBIO, flags);

#elif defined(HAVE_SETSOCKOPT_SO_NONBLOCK)

  /* BeOS */
  long b = nonblock ? 1L : 0L;
  return setsockopt(sockfd, SOL_SOCKET, SO_NONBLOCK, &b, sizeof(b));

#else
#  error "no non-blocking method was found/used/set"
#endif
}

static int default_aclose(ci_socket_t sock, void *user_data)
{
  (void)user_data;

#if defined(HAVE_CLOSESOCKET)
  return closesocket(sock);
#elif defined(HAVE_CLOSESOCKET_CAMEL)
  return CloseSocket(sock);
#elif defined(HAVE_CLOSE_S)
  return close_s(sock);
#else
  return close(sock);
#endif
}

static ci_socket_t default_asocket(int domain, int type, int protocol,
                                     void *user_data)
{
  ci_socket_t s;
  (void)user_data;

  s = socket(domain, type, protocol);
  if (s == CI_SOCKET_BAD) {
    return s;
  }

  if (setsocknonblock(s, 1) != 0) {
    goto fail; /* LCOV_EXCL_LINE */
  }

#if defined(FD_CLOEXEC) && !defined(MSDOS)
  /* Configure the socket fd as close-on-exec. */
  if (fcntl(s, F_SETFD, FD_CLOEXEC) != 0) {
    goto fail; /* LCOV_EXCL_LINE */
  }
#endif

  /* No need to emit SIGPIPE on socket errors */
#if defined(SO_NOSIGPIPE)
  {
    int opt = 1;
    (void)setsockopt(s, SOL_SOCKET, SO_NOSIGPIPE, (void *)&opt, sizeof(opt));
  }
#endif


  if (type == SOCK_STREAM) {
    int opt = 1;

#ifdef TCP_NODELAY
    /*
     * Disable the Nagle algorithm (only relevant for TCP sockets, and thus not
     * in configure_socket). In general, in DNS lookups we're pretty much
     * interested in firing off a single request and then waiting for a reply,
     * so batching isn't very interesting.
     */
    if (setsockopt(s, IPPROTO_TCP, TCP_NODELAY, (void *)&opt, sizeof(opt)) !=
        0) {
      goto fail;
    }
#endif
  }

#if defined(IPV6_V6ONLY) && defined(USE_WINSOCK)
  /* Support for IPv4-mapped IPv6 addresses.
   * Linux kernel, NetBSD, FreeBSD and Darwin: default is off;
   * Windows Vista and later: default is on;
   * DragonFly BSD: acts like off, and dummy setting;
   * OpenBSD and earlier Windows: unsupported.
   * Linux: controlled by /proc/sys/net/ipv6/bindv6only.
   */
  if (domain == PF_INET6) {
    int on = 0;
    (void)setsockopt(s, IPPROTO_IPV6, IPV6_V6ONLY, (void *)&on, sizeof(on));
  }
#endif

  return s;

fail:
  default_aclose(s, user_data);
  return CI_SOCKET_BAD;
}

static int default_asetsockopt(ci_socket_t sock, ci_socket_opt_t opt,
                               const void *val, ci_socklen_t val_size,
                               void *user_data)
{
  switch (opt) {
    case CI_SOCKET_OPT_SENDBUF_SIZE:
      if (val_size != sizeof(int)) {
        SET_SOCKERRNO(EINVAL);
        return -1;
      }
      return setsockopt(sock, SOL_SOCKET, SO_SNDBUF, val, val_size);

    case CI_SOCKET_OPT_RECVBUF_SIZE:
      if (val_size != sizeof(int)) {
        SET_SOCKERRNO(EINVAL);
        return -1;
      }
      return setsockopt(sock, SOL_SOCKET, SO_RCVBUF, val, val_size);

    case CI_SOCKET_OPT_BIND_DEVICE:
      if (!ci_str_isprint(val, (size_t)val_size)) {
        SET_SOCKERRNO(EINVAL);
        return -1;
      }
#ifdef SO_BINDTODEVICE
      return setsockopt(sock, SOL_SOCKET, SO_BINDTODEVICE, val, val_size);
#else
      SET_SOCKERRNO(ENOSYS);
      return -1;
#endif

    case CI_SOCKET_OPT_TCP_FASTOPEN:
      if (val_size != sizeof(ci_bool_t)) {
        SET_SOCKERRNO(EINVAL);
        return -1;
      }
#if defined(TFO_CLIENT_SOCKOPT)
      {
        int                oval;
        const ci_bool_t *pval = val;
        oval                    = (int)*pval;
        return setsockopt(sock, IPPROTO_TCP, TFO_CLIENT_SOCKOPT, (void *)&oval,
                          sizeof(oval));
      }
#elif TFO_SUPPORTED
      return 0;
#else
      SET_SOCKERRNO(ENOSYS);
      return -1;
#endif
  }

  (void)user_data;
  SET_SOCKERRNO(ENOSYS);
  return -1;
}

static int default_aconnect(ci_socket_t sock, const struct sockaddr *address,
                            ci_socklen_t address_len, unsigned int flags,
                            void *user_data)
{
  (void)user_data;

#if defined(TFO_SKIP_CONNECT) && TFO_SKIP_CONNECT
  if (flags & CI_SOCKET_CONN_TCP_FASTOPEN) {
    return 0;
  }
  return connect(sock, address, address_len);
#elif defined(TFO_USE_CONNECTX) && TFO_USE_CONNECTX
  if (flags & CI_SOCKET_CONN_TCP_FASTOPEN) {
    sa_endpoints_t endpoints;

    memset(&endpoints, 0, sizeof(endpoints));
    endpoints.sae_dstaddr    = address;
    endpoints.sae_dstaddrlen = address_len;

    return connectx(sock, &endpoints, SAE_ASSOCID_ANY,
                    CONNECT_DATA_IDEMPOTENT | CONNECT_RESUME_ON_READ_WRITE,
                    NULL, 0, NULL, NULL);
  } else {
    return connect(sock, address, address_len);
  }
#else
  (void)flags;
  return connect(sock, address, address_len);
#endif
}

static ci_ssize_t default_arecvfrom(ci_socket_t sock, void *buffer,
                                      size_t length, int flags,
                                      struct sockaddr *address,
                                      ci_socklen_t  *address_len,
                                      void            *user_data)
{
  (void)user_data;

#ifdef HAVE_RECVFROM
  return (ci_ssize_t)recvfrom(sock, buffer, (RECVFROM_TYPE_ARG3)length, flags,
                                address, address_len);
#else
  if (address != NULL && address_len != NULL) {
    memset(address, 0, (size_t)*address_len);
    address->sa_family = AF_UNSPEC;
  }
  return (ci_ssize_t)recv(sock, buffer, (RECVFROM_TYPE_ARG3)length, flags);
#endif
}

static ci_ssize_t default_asendto(ci_socket_t sock, const void *buffer,
                                    size_t length, int flags,
                                    const struct sockaddr *address,
                                    ci_socklen_t address_len, void *user_data)
{
  (void)user_data;

  if (address != NULL) {
#ifdef HAVE_SENDTO
    return (ci_ssize_t)sendto((SEND_TYPE_ARG1)sock, (SEND_TYPE_ARG2)buffer,
                                (SEND_TYPE_ARG3)length, (SEND_TYPE_ARG4)flags,
                                address, address_len);
#else
    (void)address_len;
#endif
  }

  return (ci_ssize_t)send((SEND_TYPE_ARG1)sock, (SEND_TYPE_ARG2)buffer,
                            (SEND_TYPE_ARG3)length, (SEND_TYPE_ARG4)flags);
}

static int default_agetsockname(ci_socket_t sock, struct sockaddr *address,
                                ci_socklen_t *address_len, void *user_data)
{
  (void)user_data;
  return getsockname(sock, address, address_len);
}

static int default_abind(ci_socket_t sock, unsigned int flags,
                         const struct sockaddr *address, socklen_t address_len,
                         void *user_data)
{
  (void)user_data;

#ifdef IP_BIND_ADDRESS_NO_PORT
  if (flags & CI_SOCKET_BIND_TCP && flags & CI_SOCKET_BIND_CLIENT) {
    int opt = 1;
    (void)setsockopt(sock, SOL_IP, IP_BIND_ADDRESS_NO_PORT, &opt, sizeof(opt));
  }
#else
  (void)flags;
#endif

  return bind(sock, address, address_len);
}

static unsigned int default_aif_nametoindex(const char *ifname, void *user_data)
{
  (void)user_data;
  return ci_os_if_nametoindex(ifname);
}

static const char *default_aif_indextoname(unsigned int ifindex,
                                           char        *ifname_buf,
                                           size_t       ifname_buf_len,
                                           void        *user_data)
{
  (void)user_data;
  return ci_os_if_indextoname(ifindex, ifname_buf, ifname_buf_len);
}

static const struct ci_socket_functions_ex default_socket_functions = {
  1,
  CI_SOCKFUNC_FLAG_NONBLOCKING,
  default_asocket,
  default_aclose,
  default_asetsockopt,
  default_aconnect,
  default_arecvfrom,
  default_asendto,
  default_agetsockname,
  default_abind,
  default_aif_nametoindex,
  default_aif_indextoname
};

void ci_set_socket_functions_def(ci_channel_t *channel)
{
  ci_set_socket_functions_ex(channel, &default_socket_functions, NULL);
}

static int legacycb_aclose(ci_socket_t sock, void *user_data)
{
  ci_channel_t *channel = user_data;

  if (channel->legacy_sock_funcs != NULL &&
      channel->legacy_sock_funcs->aclose != NULL) {
    return channel->legacy_sock_funcs->aclose(
      sock, channel->legacy_sock_funcs_cb_data);
  }

  return default_aclose(sock, NULL);
}

static ci_socket_t legacycb_asocket(int domain, int type, int protocol,
                                      void *user_data)
{
  ci_channel_t *channel = user_data;

  if (channel->legacy_sock_funcs != NULL &&
      channel->legacy_sock_funcs->asocket != NULL) {
    return channel->legacy_sock_funcs->asocket(
      domain, type, protocol, channel->legacy_sock_funcs_cb_data);
  }

  return default_asocket(domain, type, protocol, NULL);
}

static int legacycb_asetsockopt(ci_socket_t sock, ci_socket_opt_t opt,
                                const void *val, ci_socklen_t val_size,
                                void *user_data)
{
  (void)sock;
  (void)opt;
  (void)val;
  (void)val_size;
  (void)user_data;
  SET_SOCKERRNO(ENOSYS);
  return -1;
}

static int legacycb_aconnect(ci_socket_t sock, const struct sockaddr *address,
                             ci_socklen_t address_len, unsigned int flags,
                             void *user_data)
{
  ci_channel_t *channel = user_data;

  if (channel->legacy_sock_funcs != NULL &&
      channel->legacy_sock_funcs->aconnect != NULL) {
    return channel->legacy_sock_funcs->aconnect(
      sock, address, address_len, channel->legacy_sock_funcs_cb_data);
  }

  return default_aconnect(sock, address, address_len, flags, NULL);
}

static ci_ssize_t legacycb_arecvfrom(ci_socket_t sock, void *buffer,
                                       size_t length, int flags,
                                       struct sockaddr *address,
                                       ci_socklen_t  *address_len,
                                       void            *user_data)
{
  ci_channel_t *channel = user_data;

  if (channel->legacy_sock_funcs != NULL &&
      channel->legacy_sock_funcs->arecvfrom != NULL) {
    if (address != NULL && address_len != NULL) {
      memset(address, 0, (size_t)*address_len);
      address->sa_family = AF_UNSPEC;
    }
    return channel->legacy_sock_funcs->arecvfrom(
      sock, buffer, length, flags, address, address_len,
      channel->legacy_sock_funcs_cb_data);
  }

  return default_arecvfrom(sock, buffer, length, flags, address, address_len,
                           NULL);
}

static ci_ssize_t legacycb_asendto(ci_socket_t sock, const void *buffer,
                                     size_t length, int flags,
                                     const struct sockaddr *address,
                                     ci_socklen_t         address_len,
                                     void                  *user_data)
{
  ci_channel_t *channel = user_data;

  if (channel->legacy_sock_funcs != NULL &&
      channel->legacy_sock_funcs->asendv != NULL) {
    struct iovec vec;
    vec.iov_base = (void *)((size_t)buffer); /* Cast off const */
    vec.iov_len  = length;
    return channel->legacy_sock_funcs->asendv(
      sock, &vec, 1, channel->legacy_sock_funcs_cb_data);
  }

  return default_asendto(sock, buffer, length, flags, address, address_len,
                         NULL);
}


static const struct ci_socket_functions_ex legacy_socket_functions = {
  1,
  0,
  legacycb_asocket,
  legacycb_aclose,
  legacycb_asetsockopt,
  legacycb_aconnect,
  legacycb_arecvfrom,
  legacycb_asendto,
  NULL, /* agetsockname */
  NULL, /* abind */
  NULL, /* aif_nametoindex */
  NULL  /* aif_indextoname */
};

void ci_set_socket_functions(ci_channel_t                     *channel,
                               const struct ci_socket_functions *funcs,
                               void                               *data)
{
  if (channel == NULL || channel->optmask & CI_OPT_EVENT_THREAD) {
    return;
  }

  channel->legacy_sock_funcs         = funcs;
  channel->legacy_sock_funcs_cb_data = data;
  ci_set_socket_functions_ex(channel, &legacy_socket_functions, channel);
}
